### PR TITLE
Support heuristic-based scale window computation

### PR DESCRIPTION
**Pull request by cbalioglu**
This PR introduces the same heuristic method as in fairseq to determine the optimum scale window in `DynamicLossScaler`.

### CODE REVIEW

File: src/fairseq2/optim/dynamic_loss_scaler.py [hunk 1]
```diff
@@ -77,6 +83,12 @@ def __init__(
                             f"The parameters held by `optimizer` must be on a 'cuda' device, but at least one parameter is on a '{param.device.type}' device."
                         )
 
+        if scale_window is None:
+            # This is the formula that we use in fairseq.
+            scale_window = max(int(2**14 / gang.size / gradient_accumulation), 1)
```
Comment (shagunsodhani) [inline src/fairseq2/optim/dynamic_loss_scaler.py:88]: interesting! Is there any pointers to why this is good - out of curiosity :)

Comment (cbalioglu) [reply src/fairseq2/optim/dynamic_loss_scaler.py:88]: Good point :) I was also curious about that and traced it down to a fairinternal PR back from 2018 authored by Sergey. There is no mention why specifically they (Sergey and Myle) picked 2**14 and not some other value. I assume they experimented with different values and decided that this was the optimal one. To be consistent with fairseq, I just copied it over to fairseq2. I can ask Sergey next time I see him :)

Comment (shagunsodhani) [reply src/fairseq2/optim/dynamic_loss_scaler.py:88]: Thank you for the context!

Comment (cbalioglu) [reply src/fairseq2/optim/dynamic_loss_scaler.py:88]: np!

**Pull request merged**